### ISSUE TITLE
Difficulty using screen reader in "Submit an Article" page of the dashboard

### ISSUE THREAD

**Issue by benaltair** [pkp/pkp-lib#7006]
**Describe the bug**
Certain dashboard items have visual hints like `*` that are read out by the screen reader. Upon a quick inspection of the code, it looks like `aria-hidden` is not employed here when it should be. There are likely many more examples than I'll provide below.

**To Reproduce**
1. Go to `/submission/wizard` to submit a new article
2. Enable a screen reader
3. Tab through to a required field
4. Listen for "dot star" or similar

**What application are you using?**
 OJS 3.2.1.2

**Additional information**
![image](https://user-images.githubusercontent.com/17771229/116613225-44ce0500-a8ed-11eb-9771-4d732adb04a2.png)

**Comment by benaltair** [comment]
For this particular issue, the problem arises from the `<span class="req">` being read aloud.
```html
<label class="sub_label" for="locale">
	Submissions in several languages are accepted. Choose the primary language of the submission from the pulldown above.<span class="req">*</span></label>
```
Simply adjusting the last line to `<span class="req" aria-hidden="true">` should fix that issue. Since `aria-required` is already added as an attribute to the input element, it's not needed to announce this condition again.

If desired, this issue could be closed immediately after this fix. I was thinking however that it might be helpful to review the different elements with a screen reader more thoroughly. I know a11y is important to PKP and efforts have already been under way - maybe this is already taken care of.

**Comment by benaltair** [comment]
Another example on this page of a difficulty navigating with a screen reader if the "Comments for the Editor" section with the rich text editor. In my experience my screen reader seemed to fail to link the `label` to the `textarea`.

```html
<div class="section">
  <label for="commentsToEditor-608b147fe61ce">Comments for the Editor</label>
  <div>
  <!-- control markup removed for brevity... -->
          <iframe
            id="commentsToEditor-608b147fe61ce_ifr"
            frameborder="0"
            allowtransparency="true"
            title="Rich Text Area. Press ALT-F9 for menu. Press ALT-F10 for toolbar. Press ALT-0 for help"
            style="width: 100%; height: 200px; display: block"
          ></iframe>
        </div>
      </div>
    </div>
    <textarea
      class="richContent"
      name="commentsToEditor"
      rows="10"
      cols="80"
      id="commentsToEditor-608b147fe61ce"
      style="display: none"
      aria-hidden="true"
    ></textarea>
    <span></span>
  </div>
</div>
```

I think this is actually due to the `textarea` not being used. Instead it seems that a `contenteditable` element within the `iframe` is actually being selected by the user.

```html
<!-- within tinymce iframe -->
<body id="tinymce" class="mce-content-body " data-id="commentsToEditor-608b147fe61ce" contenteditable="true" spellcheck="false"><p><br data-mce-bogus="1"></p></body>
```

Since this field is not semantically linked to the label, the experience with a screen reader is not ideal. Perhaps one solution would be to add an `id` to the `label` element:
```html
<label for="commentsToEditor-608b147fe61ce" id="comments-for-editor">Comments for the Editor</label>
```
Then add `aria-labelledby` to the div containing the iframe (thus preserving the help-text added to the iframe itself).
```html
<div id="mceu_27" class="mce-edit-area mce-container mce-panel mce-stack-layout-item mce-last" hidefocus="1" tabindex="-1" role="group" style="border-width: 1px 0px 0px;" aria-labelledby="comments-for-editor">
<iframe id="commentsToEditor-608b147fe61ce_ifr" frameborder="0" allowtransparency="true" title="Rich Text Area. Press ALT-F9 for menu. Press ALT-F10 for toolbar. Press ALT-0 for help" style="width: 100%; height: 200px; display: block;"></iframe>
</div>
```
As you can see above, `aria-labelledby="comments-for-editor"` semantically links the grouping (denoted by `role="group"` which is already present) to the label above. Now when you enter the text editor application using a screen reader, it announces that you've entered the "Comments for the editor" group, thus fixing the problem.

**Comment by NateWr** [comment]
Thanks for testing, @benaltair! We're rebuilding the whole submission wizard in 3.4 to use the new forms. Our plan is to then have it audited by an accessibility expert. 

Would you be able to do similar tests for us on the settings forms? A couple examples are Settings > Journal > Masthead, Settings > Journal > Contact. These are the new forms that we'll be using for the submission wizard, and which we plan to roll out to every part of our application.

**Comment by benaltair** [comment]
Did a quick review and saw some small issues. I'm using NVDA as the reader, one of the most widely used free options.

## Multi-line Text Editor
The experience compared with the other forms is much better. For just entering plain text it was better than the documented experience above.
![image](https://user-images.githubusercontent.com/17771229/117179771-53eb0200-ad88-11eb-9e03-a0e9ba61d210.png)
Just two issues seem to come up upon initial testing.
### 1. The toolbar isn't accessible with a keyboard
The MCE editor has `tabindex="-1"` scattered around various elements. Additionally `hidefocus="1"` is used in certain places. The issue is that these, as the names suggests, prevent some/most keyboard users from using the format toolbar. It's a tricky issue to solve for. For a case study, take a close look at how Github handles the tab behaviour of their inline multi-line text editors, and reverse engineer some of the code for the toolbars. I'm not sure how much flexibility you have here in terms of improving the TinyMCE library. Happy to look at this more if you have specific questions.

A goal here would be to allow a keyboard user to enter the toolbar if they wanted, or to easily skip past it and enter the main text field. You might be able to use a mechanism like [groups](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_group_role) to achieve this.

I'll mention here that in any forms you have control over, using [`fieldset`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset) is a nice way to group form components semantically (and visually if you want).

### 2. Language completion icon
The globe icon which shows you how many languages you've completed doesn't work well with a screen reader. I'll write more about that in the next section.


## Short text field
The language icon reads as **blank** with a screen reader.
![image](https://user-images.githubusercontent.com/17771229/117176954-4f711a00-ad85-11eb-988c-00c337a280ec.png)
Looking at the code, I see that `aria-hidden` is used on the `button` element (and its child). Since any `button` is tab selectable, and readers will mostly read what is selected using the tab key, this attribute doesn't have any effect for user interaction. `aria-hidden` as I understand it, can't be used on tab-selectable elements.
```html
<span class="multilingualProgress multilingualProgress--isEmpty" id="masthead-abbreviation-multilingualProgress-en_US">
    <button type="button" aria-hidden="true" data-original-title="null" class=" has-tooltip">
      <span aria-hidden="true" class="fa fa-globe"></span>
    </button>
  <span class="-screenReader">0/3 languages completed</span>
</span>
```
Additionally, there is a screen-reader-specific `span` afterwards. While this is common practice, it might be worth using the more semantic `aria-label` or `aria-labelledby` attributes on the `button` element itself. This way as the user is tabbing through a page, they'll get contextual verbal feedback. These attributes are widely supported by now as I understand, with all major readers treating them almost identically. In fact, from what I've heard, the hidden `span` actually will be treated more inconsistently, for instance between mobile/desktop versions of a reader.

I'll also mention that assistive labels would be better typed out fully, rather than using symbols like `/`. For instance, `0/3` could be read by some as `zero thirds`, `zero divided by 3`, `zero slash three` or `zero three`. Instead using `0 of 3 languages completed` in the label would keep things consistent across readers.

It's possible that a `button` element here doesn't work, as buttons should trigger an action (and `a` links should trigger a navigation). It's sort of a grey area, since visually they toggle a tooltip, while for unsighted users one could imagine it would be redundant to have the following steps:
1. Tab to button which reads "show number of languages currently completed"
2. Toggle button
3. Reads "0 of 3 languages completed"

An example of how these language buttons could be handled is:
```html
<span aria-label="0 of 3 languages completed" class="multilingualProgress multilingualProgress--isEmpty" id="masthead-abbreviation-multilingualProgress-en_US">
    <span aria-hidden="true" class="fa fa-globe"></span>
</span>
```
Of course this would need complimentary changes to JS and CSS to enable the tooltip (which could be on hover). An alternative which may be semantically simpler is to have a hidden tooltip element within that is only visible when the icon is hovered over, and will also be read by a screen reader.

**Comment by NateWr** [comment]
Thanks for this great report, @benaltair! It's really helpful as we prepare to get our backend UI audited. On TinyMCE, it does claim to be [compatible with JAWS and NVDA](https://www.tiny.cloud/docs/advanced/accessibility/). Would you have time to read that and see if the shortcuts are working properly?

Thanks for the tip about `aria-hidden` not working on clickable elements. I didn't know that! If `aria-hidden` is used on a parent element, will it hide child elements like this:

```
<span aria-hidden="true">
  <button></button>
</span>
```

Or will the `<button>` still be accessible to a screen reader?

Finally, can you provide some more detail about the `<fieldset>`/`<legend>` issue? I would expect a screen reader to be able to read the legend without requiring an additional `aria-labelledby` attribute. Is the problem that the description is not accessible?

I think that the `<legend>` has to be a direct child of `<fieldset>` (see [here](https://www.accessibility-developer-guide.com/examples/forms/grouping-with-fieldset-legend/#legend-must-be-a-direct-child-of-fieldset) and [here](https://stackoverflow.com/questions/44202283/fieldsets-legend-inside-of-a-div-fails-pa11y-wcag-2-0)). But I think it's possible to use `aria-labelledby` with multiple IDs, so we could do:

```
<fieldset aria-labelledby="legend description">
  <legend id="legend">...</legend>
  <div id="description">...</div>
</fieldset>
```

But is that right? It feels wrong to override the default pairing of `<fieldset>`/`<legend>` from the HTML spec.

**Comment by benaltair** [comment]
# TinyMCE A11y
Hi @NateWr, my pleasure. I checked out the keyboard shortcuts from TinyMCE and they do mention Alt+F10 for the toolbar but afaik that is already used by browsers like Chrome to select _their_ toolbar. I couldn't get it to work for me. Additionally, I've seen an `aria-label` attribute used on a parent of a widget (like a multiline text editor) to announce available keyboard shortcuts. That might be one way to tackle this, but as I mentioned I'm not sure why it wasn't working for me. Frankly, I don't think TinyMCE has designed this intuitively. 
![image](https://user-images.githubusercontent.com/17771229/117490989-72860000-af24-11eb-88d5-eb34d64f8078.png)

# `aria-hidden` with focusable child elements
On the question about child elements with `aria-hidden`, this excerpt from MDN is very helpful:

> According to the [fourth rule of ARIA](https://www.w3.org/TR/using-aria/#fourth), `aria-hidden="true"` should not be used on a focusable element. Additionally, since this attribute is inherited by an element's children, it should not be added onto the parent or ancestor of a focusable element. 
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-hidden_attribute (draft)

Short answer is no from what I understand. In my experience `aria-hidden` will not be passed down to a focusable child. The `<button>` would still be accessible with a screen reader.

# `fieldset` and `legend`
Great catch about `legend` needing to be the initial child of `fieldset`. Upon further investigation, it doesn't seem like this is the case on either Masthead or Contact tabs. In both tabs, `fieldset` elements have `legend` elements, but they aren't direct children. A quick modification to the HTML in dev tools to make `legend` the first child of `fieldset` fixed the issue, and the grouping was announced on tab selection.
![image](https://user-images.githubusercontent.com/17771229/117491975-d3fa9e80-af25-11eb-9a15-3d073ecfbb79.png)
So your example you quoted would be perfect and then `aria-label` wouldn't be needed. The issue here is that since `legend` isn't a direct child of `fieldset` (at least on my version of OJS), the announcement is not made.

**Comment by benaltair** [comment]
@NateWr I hope I'm wrong about TinyMCE and just missing something - I'm sure the audits will help figure that out 😊

The issues look good. I'll add more details if something comes to mind.

**Comment by benaltair** [comment]
Hi @NateWr, just realized I closed this issue when there were still a number of other things that could be addressed. I think I meant to close my repo's issue on the same topic. 

Did you want to keep this closed if they're being addressed elsewhere or should we open it?

**Comment by NateWr** [comment]
@benaltair I'd say let's keep this closed and wait for our accessibility audit on the new submission wizard (hopefully in 2022). Is there anything remaining that is not captured in #7045 and #7046? If so, let's file a new issue for that.

**Comment by benaltair** [comment]
Yes, sounds good - on first glance they cover everything @NateWr
